compiler/rustc_hir_typeck/src/method/mod.rs

```diff
@@ -276,6 +283,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         method_name: Ident,
         self_ty: Ty<'tcx>,
         call_expr: &hir::Expr<'_>,
+        expected_def_id: Option<DefId>,
         scope: ProbeScope,
```

Comment:
I really don't think we need to add another argument to `lookup_probe` -- This can be consolidated into a `ProbeScope` variant. Like `ProbeScope::Single` or something like that.

Reply:
We are still using the `ProbeScope::TraitsInScope` logic for walking through the candidates and building the deref/coercion chain,  just filter away candidates not matching the expected `DefId`.
Technically, `Some(expected_def_id)` is compatible with `ProbeScope::AllTraits` too, it's just not used at the moment.

So these things are orthogonal.

Reply:
I'm not certain why that is orthogonal. Why should we build a bunch of candidates then filter after the fact if we know the exact candidate that we should probe? 

Auto(de)ref shouldn't really matter here.

Reply:
Let me check again.

Reply:
Do you mean skipping the whole `probe_cx.assemble_extension_candidates_for_traits_in_scope()` logic and make a set out of a single candidate?
Upd: we'll need to skip `probe_cx.assemble_inherent_candidates()` too.

After that we won't even need to pass the `expected_def_id` because the only candidate will already have that id by definition.
Yeah, that should work I guess.

Reply:
Extending the set of traits in scope in `rustc_resolve` should also no longer be necessary if we do as the comment above says.